Transition dipole moments from DeltaDFTB

timkowalczyk opened the pull request:
Implements evaluation of the transition dipole moment between DFTB ground state and DeltaDFTB excited state as described in Deshaye et al. JCP 158, 134104. Adds a regression test for the feature and updates the manual.

Commit:
WIP Add TI transition dipole moments from dev ver of 19.1

Fixes to allow code to run

Summary:
* Use of unitialised sizing variables fixed (nLocal* were only set
  for the hybrid functional code path in initprogram.F90, and na
  was never set in determinants.F90)
* Mulliken return sizing incompatible with variable being used, so
  introduce work array to match these together
* tiTraCharges was zeroing all spin channels when it should only
  zero the current one in determinants
* Fix for hamiltonian being stored after overwrite with
  eigenvectors

Further fixes now reproduce TI-DFTB TDM for CO

* Initialize corresponding orbital coefficient matrices
  to zero at each build to resolve NaNs in TDMs
* Several corrections to iDeterminant conditionals to
  ensure proper construction of gfilling and mfilling
* Removed TDMs from writeDetailedOut7 since they are not
  built in time for this subroutine; print directly for now

Clean up, add reg test, and document TDM feature

* Remove unused iterators and obsolete comments

* Clean up formatting of write() for TDMs to match
  formatting for permanent dipole moments

* Add transitionDipoleMoment to taggable properties
  in autotest

* Add regression test for TDM of CO

* Document the new TDM logical and its use in the
  DeltaDFTB subsection of the manual. Includes one
  new literature reference.

bhourahine commented on src/dftbp/dftb/determinants.F90:
```diff
@@ -471,5 +496,228 @@ pure subroutine applyZieglerAlloc(eM, eT, eF)
 
   end subroutine applyZieglerAlloc
 
+  subroutine tiTDM(g, e, pt, gfilling, mfilling)
```
```suggestion
  !> Evaluate transition density matrix for time independent excitation
  subroutine tiTDM(g, e, pt, gfilling, mfilling)
```

bhourahine commented on src/dftbp/dftb/determinants.F90:
```diff
@@ -289,6 +310,9 @@ subroutine TDftbDeterminants_init(this, isNonAufbau, isSpinPurify, isGroundGuess
     !> Should there be a ground state initial guess before Non-Aufbau calc?
     logical, intent(in) :: isGroundGuess
 
+    !> Should the transition dipole moment with the ground state be computed?
```
```suggestion
    !> Should the transition dipole moment between the excited and ground state be computed?
```

bhourahine reviewed (commented):
Just some comments, but also, what happens for periodic systems and MPI enabled calculations? We might need to block those cases for the moment.

bhourahine committed:
Trap for MPI compiled code

Remove language depreciated forall, fix some indentation and
rename a couple of variables as per style guide. Also switch
initialisation to current error return mechanism (needed
eventualy for library use of DFTB+).

Also switch to more recent mechanism for finding SK files, removing
need for symlinked files in the new test.

Also enabled MPI running of Delta-DFTB cases where possible.
Block periodic geometries for transition dipole
